### ISSUE TITLE
Create test pattern for Spans

### ISSUE THREAD

**Issue by alamb** [apache/datafusion-sqlparser-rs#1563]
Part of https://github.com/apache/datafusion-sqlparser-rs/issues/1548

# Usecase

We have added span information in https://github.com/apache/datafusion-sqlparser-rs/pull/1435

We would like to add span information through out the rest of the AST structures (see docs in https://github.com/apache/datafusion-sqlparser-rs/pull/1549) and list on https://github.com/apache/datafusion-sqlparser-rs/issues/1548

In order to support this feature reasonably, we need to be able to test the feature and avoid regressions

# Proposal

I think we need some sort of span test.

I recommend
1. A new test binary like `tests/sqlparser_spans.rs`

Then add a test that
1. Parses a SQL string
2. Check the spans of the AST nodes match what is expected

I think we'll have to use judgement on how fine grained the location information
can / should be.

**Comment by graup** [comment]
Hi! Just a small note since I started playing around with this.  I assume it's because it's not implemented yet but anyway to keep track: it seems currently that Function args don't have a span, and also the Function span only covers the function name, not its arguments. So for `CURRENT_DATE('UTC')`, the ast node for the Function expr has a span containing only `CURRENT_DATE` (and func.args has an empty span). I'd expect the function expr span to cover the whole expression including the parentheses.

If I can help in anyway (like writing a test case), I'd be happy to contribute.

Very tangential, it would be useful if there was a helper function to convert line-column-based locations to byte offsets. (My use case is implementing a fixer without rebuilding the SQL from the AST, just applying fixed locally to individual nodes.)

Input: `SELECT CURRENT_DATE(),\nCURRENT_DATE('UTC')`
Output: `
[Query(Query { with: None, body: Select(Select { select_token: TokenWithSpan { token: Word(Word { value: "SELECT", quote_style: None, keyword: SELECT }), span: Span(Location(1,1)..Location(1,7)) }, distinct: None, top: None, top_before_distinct: false, projection: [UnnamedExpr(Function(Function { name: ObjectName([Ident { value: "CURRENT_DATE", quote_style: None, span: Span(Location(1,8)..Location(1,20)) }]), uses_odbc_syntax: false, parameters: None, args: List(FunctionArgumentList { duplicate_treatment: None, args: [], clauses: [] }), filter: None, null_treatment: None, over: None, within_group: [] })), UnnamedExpr(Function(Function { name: ObjectName([Ident { value: "CURRENT_DATE", quote_style: None, span: Span(Location(2,1)..Location(2,13)) }]), uses_odbc_syntax: false, parameters: None, args: List(FunctionArgumentList { duplicate_treatment: None, args: [Unnamed(Expr(Value(SingleQuotedString("UTC"))))], clauses: [] }), filter: None, null_treatment: None, over: None, within_group: [] }))], into: None, from: [], lateral_views: [], prewhere: None, selection: None, group_by: Expressions([], []), cluster_by: [], distribute_by: [], sort_by: [], having: None, named_window: [], qualify: None, window_before_qualify: false, value_table_mode: None, connect_by: None }), order_by: None, limit: None, limit_by: [], offset: None, fetch: None, locks: [], for_clause: None, settings: None, format_clause: None })]
`

**Comment by lustefaniak** [comment]
Hi @alamb, what I did downstream couldn't be ported immediately as it is a much higher level testing, which I think might not work well in the low-level library. I work on the result of the AST analysis, storing locations into higher-level concepts like columns and dependency chains. Categorising locations into specific categories:
```
  SOURCE_LOCATION_KIND_PROJECTION,
  SOURCE_LOCATION_KIND_IDENTIFIER,
  SOURCE_LOCATION_KIND_EXPRESSION,
  SOURCE_LOCATION_KIND_SELECTION etc.
```

I assert that the captured "location" is precisely a specific string.

```
    assert_location_contains(
        sql,
        &get_column_dep_locations(&res, "t", "started_at"),
        &snip(LocationKind::Identifier, "started_at"),
    );
    assert_location_contains(
        sql,
        &get_column_dep_locations(&res, "t", "started_at"),
        &snip(
            LocationKind::Expression,
            "GREATEST(finished_at, started_at)",
        ),
    );
    assert_location_contains(
        sql,
        &get_column_dep_locations(&res, "t", "started_at"),
        &snip(
            LocationKind::Projection,
            "GREATEST(finished_at, started_at) as t",
        ),
    );
```

On the one hand, it is a very high level without looking into details of all the AST nodes; on the other hand, I found working with line/char super hard to grasp and error-prone and led to not the level of testing I wanted. Ultimately, I wanted the best experience for our users, and snipped-based testing worked pretty well.

**Comment by Nyrox** [comment]
> /edit to add: I guess it's because `Spanned for Value` is not implemented yet. But even if it were, I'm not sure the function span will cover all the text until the closing brace, since it's just union'ing the spans of the parsed elements, which wouldn't include the braces...? I have a feeling `Spanned for FunctionArguments` needs to be extended to the open and closing braces in the List and Subquery case. [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/7643957?s=64&amp;u=f6bc582a61a82d2f35134121d42beef484bd2162&amp;v=4">@Nyrox](https://github.com/Nyrox) wdyt?

@graup Yes you are right. There is a lot of missing cases like these. For FunctionArgumentList, adding the parenthesis as [AttachedToken](https://docs.rs/sqlparser/0.53.0/sqlparser/ast/helpers/attached_token/struct.AttachedToken.html) would be straight forward and should be done. I never quite figured out what the best to do it for enum tuple variants like Subquery is though.

> So for CURRENT_DATE('UTC'), the ast node for the Function expr has a span containing only CURRENT_DATE (and func.args has an empty span). I'd expect the function expr span to cover the whole expression including the parentheses.

Yeah this is definitely the Span for Value missing. Getting that in is pretty high priority imo.
Not sure if the best way here is to actually implemented Spanned for every Value or to wrap every usage of value in a WithSpan<T> like wrapper. Both are unfortunately quite breaking changes.
